onnx/namespace.proto

```diff
@@ -0,0 +1,55 @@
+// Copyright (c) Facebook Inc. and Microsoft Corporation.
+// Licensed under the MIT license.
+
+syntax = "proto2";
+
+package onnx;
+
+// This protobuf defines the ONNX namespace metadata, which is used to
+// resolve operator names (op_type fields in onnx.proto) into operator
+// definitions (as defined in onnx/defs).
+//
+// Semantically, ChangelogProto defines the following mapping:
+//
+//    { op_version: { op_type: op_def } }
+//
+// i.e., for every operator namespace version, a mapping of operator
+// names to operator definitions.  However, in the protobuf, this is
+// represented in a compressed, 'changelog' style format.
+// You can compute the underlying mapping with the following algorithm:
+//
+//  - Let M be an empty mapping
+//  - For each changelog entry
+//      - Remove all 'removed' keys from M
+//      - Add all 'added' entries to M
+//          (it is an error if an entry to be added already exists)
+//      - Yield M as the mapping for 'op_version'
+
+message ChangelogProto {
+  repeated ChangelogEntryProto = 1;
+}
+
+message ChangelogEntryProto {
+  // The operator namespace version this changelog entry describes.
+  optional int64 op_version = 1;
+
+  // The operator name bindings which were removed in this version.
+  // These removals are backwards compatibility BREAKING changes; so
+  // whenever you remove a name, you must issue a new version of the
+  // operator namespace.
+  repeated string removed = 2; // namespace Operator
+
+  // The operator name bindings which were added in this version.
+  // Extra name bindings can be added to this field in a backwards
+  // compatible way without bumping the operator namespace version.
+  repeated OperatorBindingProto added = 3;
+}
```

Comment:
question please. For semantic changes of an existing operator (say, Slice you mentioned :) ), will we only have it in "added" (overwrite way) or we also want to have it firstly in "removed" and then "added"?

Reply:
Yep. So let's say you have `_OldSlice` and `_NewSlice`. Going in, you'll have a version of ONNX with `Slice` mapped to `_OldSlice`. Then, to move to NewSlice, you write this changelog entry:

```
ChangelogEntry {
  removed = "Slice"
  added = OperatorBinding {
    op_type = "Slice",
    op_def = "_NewSlice",
  }
}
```

We could introduce a modify field if this is a little oblique.

Reply:
We may just have a "new" to contain both "added" and "changed". Say
ChangelogEntry {
  new = OperatorBinding {
    op_type = "Slice",
    op_def = "_NewSlice",
  }
}

In the "new", the key is always over-written, if the key exists, then it's removing the old one and change to new one, if the key does not exist, it's purely adding a new one, and "removed" contains keys totally removed.

Reply:
the name "new" may be not good although :)

Reply:
That's OK by me! My only wondering is whether or not we should be more explicit if an override occurs. "Adding" is a BC-compatible change, but "overriding" is not.